src/coprocessor/cache.rs

```diff
@@ -0,0 +1,487 @@
+                // In some coner case delete self.regions item may cause
+                // cache no need data and then cause an error result.
+                node.cached_items.remove(k);
+            }
+            self.size -= entry.size;
+        }
+    }
+
+    fn evict_region_with_exist_region(&mut self, region_id: u64) {
+        self.regions
+            .get_mut(&region_id)
+            .and_then(|region| {
+                region.version += 1;
+                Some(region.cached_items.clone())
+            })
+            .and_then(|cached_items| {
+                for (key, _) in (&cached_items).iter() {
+                    self.remove(key);
+                }
+                Some(())
+            });
+    }
+
+    fn evict_region_with_new_region(&mut self, region_id: u64) {
+        // If there's no region version data in cache, get_version will
+        // return 0 and after that some thread call evict_region, we should set
+        // version to 1 to prevent cache unused data.
+        let entry = RegionDistSQLCacheEntry {
+            version: 1,
+            ..Default::default()
+        };
+        self.regions.insert(region_id, entry);
+    }
+
+    pub fn evict_region(&mut self, region_id: u64) {
+        if self.regions.contains_key(&region_id) {
+            self.evict_region_with_exist_region(region_id);
+        } else {
+            self.evict_region_with_new_region(region_id);
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    pub fn evict_region_and_enable(&mut self, region_id: u64) {
+        self.evict_region(region_id);
+        self.enable_region_cache(region_id);
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.max_size
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.len() == 0
+    }
+
+    pub fn update_capacity(&mut self, capacity: usize) {
+        self.max_size = capacity;
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+    }
+
+    pub fn disable_region_cache(&mut self, region_id: u64) {
+        let mut entry = self.regions
+            .entry(region_id)
+            .or_insert_with(Default::default);
+        entry.enable = false;
+    }
+
+    pub fn enable_region_cache(&mut self, region_id: u64) {
+        self.regions.get_mut(&region_id).and_then(|e| {
+            e.enable = true;
+            Some(())
+        });
+    }
+
+    fn is_region_cache_enabled(&self, region_id: u64) -> bool {
+        self.regions.get(&region_id).map_or(true, |e| e.enable)
```

Comment:
Could we use a HashSet to store those ids of regions with the leader on current server? It seems we can not make sure if the leader is still on the current server when self.regions does not contain current region?

Reply:
Region's cache by default is enabled. So we do not need to track which region we have. But DistSQLObserver will keep tracking which Region is in writing. So we just track which region is writing is OK.